### PR TITLE
Dairy farm

### PR DESCRIPTION
**Pull request by AlecWhite**
First time making a location and first time making a multi-overmap-tile location. I don't understand why some fields are for and haven't been able to search information about it, like the `"sym"` in overmap_terrain and some other small problems.

What this PR adds:

- Add multi-overmap-tiles diary farm location.
- Diary farm is surrounded by a chain fence with unlocked metal doors.
- One milking barrack with 2 new terrains, bulk_tank and milking_machine.
- 20 cows.
- A small house with some loot
- Chance of spawning a cube van, a quad bike and a bicycle.
- Chance of spawning diary farm are 1 in 4
- Diary farm spawns at roughly the same location as public works

I'm awfully bad at designing houses, specially playing Sims, and it reflects it so here. I'm bad too at deciding where and what kind of loot and quantity to place. 
So any full list of where to put what kind of loot with what chance is more than welcome.

Also good lord what a PITA is to make locations in this game. When I finished I was close to sewerside.

### CODE REVIEW

**Comment by AlecWhite** [comment]
>it's dairy farm, not diary farm.

And that's why you shouldn't let dyslexic folks like me around anything that has letters in it. 

Gonna fix it.

File: data/json/overmap_terrain.json [hunk 1]
```diff
@@ -6862,5 +6862,41 @@
...
+    "id": "diary_farm_NE",
+    "type": "overmap_terrain",
+    "name": "dairy farm",
+    "sym": 119,
+    "color": "brown",
+    "see_cost": 1,
+    "flags": [ "SIDEWALK", "NO_ROTATE" ]
+  },
+  {
+    "id": "diary_farm_SE",
```
Comment (Leland) [inline data/json/overmap_terrain.json]: You got most of them – this the two above also need to be `dairy_farm`

**Comment by Leland** [comment]
Um
screen shot 2017-05-30 at 11 06 39 pm

**Comment by AlecWhite** [comment]
Uhrm, that's weird, I have it with with the overmap_special:
`"overmaps" : [
            { "point":[0,0,0], "overmap": "diary_farm_NW"},
            { "point":[1,0,0], "overmap": "diary_farm_NE"},
            { "point":[0,1,0], "overmap": "diary_farm_SW"},
            { "point":[1,1,0], "overmap": "diary_farm_SE"}
       ],`

Which is the same I have for the mod version I pasted on the PR, guess I shouldn't work on it as a mod. Gonna see why 
ss 2017-05-31 at 05 11 41

**Comment by AlecWhite** [comment]
Tried to compile the lastest version and it fails with `1>LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library`

Switched to the milking branch that is using an older version and it compiles fine. It will take me more time to see why everything has decided to go wrong now.

File: data/json/overmap_terrain.json [hunk 2]
```diff
@@ -6862,5 +6862,41 @@
...
+    "flags": [ "SIDEWALK", "NO_ROTATE" ]
+  },
+  {
+    "id": "dairy_farm_SW",
+    "type": "overmap_terrain",
+    "name": "dairy farm",
+    "sym": 119,
+    "color": "brown",
+    "see_cost": 5,
+    "flags": [ "SIDEWALK", "NO_ROTATE" ]
```
Comment (John-Candlebury) [inline data/json/overmap_terrain.json]: If I recall correctly, the weird rotation is the result of not specifying the "rotate": true field in these definitions

Comment (AlecWhite) [inline data/json/overmap_terrain.json]: Thanks will try that when if I manage to see why I can't compile.

**Comment by pisskop** [comment]
20 cows is a lot of cow and meat.

Even 5 days after groundZ Id expect most to be 'gone', be that to greener pastures or to the afterlife.

Also, no Zow?

Maybe adding a spawn group with a chance for cow would allow for 0 - x cows to spawn, and allow for a Zow to be added later.

**Comment by Coolthulhu** [comment]
> Maybe adding a spawn group with a chance for cow would allow for 0 - x cows to spawn, and allow for a Zow to be added later.

Monster group sounds good here.
At first (first few days) some of the cows could be alive, but after a season most of them should be gone, dead or converted.
It would give players a reason to "rush" the farm too.

You could also add a spawn of withered plants to the barn, to represent hay. More for flavor than anything, but it would explain why didn't the cows starve after being locked up in the way-too-small pasture.

**Comment by AlecWhite** [comment]
So what would be a good amount and chances to spawn cows?

**Comment by Coolthulhu** [comment]
> So what would be a good amount and chances to spawn cows?

Try something like ~10 at start (start is day 5, not day 0), down to ~5 by the end of first week, then near 1 by the end of second one.
Check out monstergroups.json for an example of timed groups. `starts` and `ends` are in hours.

> Also good lord what a PITA is to make locations in this game.

There was a mapmaker util at one point, I think.

File: data/json/mapgen/dairy_farm.json [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        ".$................... ..",
+        ".$................... ..",
+        ".$................... ..",
+        ".$..................... ",
+        ".$......................",
+        ".$......................",
+        ".$......................"
+      ],
+      "terrain": {
+        " ": "t_grass",
```
Comment (Coolthulhu) [inline data/json/mapgen/dairy_farm.json]: Why are some grasses spaces and other dots?

File: data/json/mapgen/dairy_farm.json [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        "$": "t_chainfence_v",
+        ".": "t_grass"
```
Comment (Coolthulhu) [inline data/json/mapgen/dairy_farm.json]: Grass should be trampled/eaten. You can add chances for dirt like this:
```
".": ["t_grass", "t_grass", "t_dirt", "t_dirt", "t_dirt"]
```
The above would give a 60% chance for dirt, 40% chance for grass.

File: data/json/mapgen/dairy_farm.json [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        { "monster": "mon_cow", "x": 19, "y": 6 },
+        { "monster": "mon_cow", "x": 9, "y": 10 },
+        { "monster": "mon_cow", "x": 11, "y": 10 },
+        { "monster": "mon_cow", "x": 17, "y": 11 },
+        { "monster": "mon_cow", "x": 7, "y": 12 },
+        { "monster": "mon_cow", "x": 9, "y": 14 },
+        { "monster": "mon_cow", "x": 13, "y": 16 },
+        { "monster": "mon_cow", "x": 6, "y": 17 },
+        { "monster": "mon_cow", "x": 8, "y": 18 },
+        { "monster": "mon_cow", "x": 13, "y": 18 }
```
Comment (Coolthulhu) [inline data/json/mapgen/dairy_farm.json]: Don't place each monster one by one, place a group like this:
```
{ "monster": "GROUP_COWS", "x": [ 1, 22 ], "y": [ 1, 22 ], "density": 0.05 }
```
Then define the cow group somewhere on top of the file.

**Comment by AlecWhite** [comment]
>There was a mapmaker util at one point, I think.

That's the one I'm using to do all the heavy lifting. It still needs to be manually tweaked and fixed here and then. Like the NO_ROTATE and "rotate" fields.

Gonna have to fix stuff on a mod due to https://github.com/CleverRaven/Cataclysm-DDA/issues/21161#issuecomment-305827208.

**Comment by AlecWhite** [comment]
@Coolthulhu What's the correct use of monster group? The linting tools gives errors due to https://github.com/CleverRaven/Cataclysm-DDA/issues/21181.

I'm using right now `"monsters":[
{"monster" : "mon_cow", "cost_multiplier" : 10, "freq": 100,  "ends" : 168, "pack_size" : [5,10] },
{"monster" : "mon_cow", "freq": 100, "cost_multiplier" : 5, "starts" :  336, "pack_size" : [1,5] }]`

I assume that "ends" means when that monster won't be used and "starts" when that monster will be used, and "pack_size" that it will be spawned in a group. So far it doesn't seems to be working.

**Comment by Coolthulhu** [comment]
> So far it doesn't seems to be working.

What is happening? No spawns at all? Pack size not decreasing?

**Comment by AlecWhite** [comment]
No cows spawned, absolutely none. Should I try removing pack_size and position cows one by one?

**Comment by Coolthulhu** [comment]
No, remove pack size but position them over an area, like in my example:
```
{ "monster": "GROUP_COWS", "x": [ 1, 22 ], "y": [ 1, 22 ], "density": 0.05 }
```

**Comment by AlecWhite** [comment]
@Coolthulhu did I place the monsters correctly?

Also would lint, but the linting web tool still throws errors.

File: data/json/mapgen/dairy_farm.json [hunk 6]
```diff
@@ -0,0 +1,251 @@
+[
+  {
+    "type":"monstergroup",
+    "name" : "GROUP_COWS",
+    "is_safe" : true,
+    "default" : "mon_null",
+    "monsters":[{"monster" : "mon_cow", "cost_multiplier" : 10, "freq": 100, "ends" : 168 },
+    {"monster" : "mon_cow", "freq": 50, "cost_multiplier" : 5, "starts" :  336  }
+    ]
```
Comment (pisskop) [inline data/json/mapgen/dairy_farm.json]: Theres a period of time where no cows at all will spawn:  between 168 and 336. 
  Im not sure if the quantity means hours or not, but theres a gap nontheless.

additionally, with a cost multiplier that high, its no wonder you wont get any cows out of a passive building.  Later on in the game theres more 'spawn points' available, but early game theres not nearly enough in your average mapgen structure to make a cow.

Try a cost of 1 or 0 for now.  You have already made them only spawn 10% of the time and 5%, respectively.

File: data/json/mapgen/dairy_farm.json [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        "furniture": {
+            "$": "f_null",
+            ".": "f_null"
+        },
+        "place_monsters": [
+        { "monster": "GROUP_COWS", "x": [ 1, 10 ], "y": [ 1, 9 ], "density": 0.5 },
+        { "monster": "GROUP_COWS", "x": [ 1, 12 ], "y": [ 1, 12 ], "density": 0.5 },
+        { "monster": "GROUP_COWS", "x": [ 1, 11 ], "y": [ 1, 10 ], "density": 0.5 },
+        { "monster": "GROUP_COWS", "x": [ 1, 9 ], "y": [ 1, 11 ], "density": 0.5 },
+        { "monster": "GROUP_COWS", "x": [ 1, 15 ], "y": [ 1, 12 ], "density": 0.5 }
```
Comment (pisskop) [inline data/json/mapgen/dairy_farm.json]: 0.5 density only hurts the cows more.

I dont personally know the exact effect density has on mons, but setting it to 1 or 5 definitely highlights that it does have some effect.  Leave density out alltogether and/or specify it in the terrain.json entry?

File: data/json/mapgen/dairy_farm.json (hunk 6 above)
Comment (AlecWhite) [inline data/json/mapgen/dairy_farm.json]: Fixed the time gap.

Didn't know spawn cost was that costly. Also what do you mean by 10? and 5%? should they spawn 100% and 50% of the time? I thought it was tied to freq.

File: data/json/mapgen/dairy_farm.json (hunk 7 above)
Comment (AlecWhite) [inline data/json/mapgen/dairy_farm.json]: I don't know what density does, just copypasting what @Coolthulhu recommended.

File: data/json/mapgen/dairy_farm.json (hunk 6 above)
Comment (pisskop) [inline data/json/mapgen/dairy_farm.json]: Spawn frequency is rate on a 1000 point scale, not 100.

The default mon, mon_null is chosen for all cases wherein a one-thousand sided die is rolled and 1-100 is  not chosen.  thus 100 points of freq is 10%

**Comment by AlecWhite** [comment]
It seems some VS stuff got in the way. Time to see how to fix it.

**Comment by AlecWhite** [comment]
I think I fixed it, have some untracked files but in theory it shouldn't be too important.

**Comment by narc0tiq** [comment]
> The linting tools gives errors due to #21181.

No, actually, the linting tool in the Jenkins workspace is as up to date as possible. The referenced issue is a reminder for me to make sure that the Web linter (at <http://dev.narc.ro/cataclysm/format.html>) continues running the exact same code as the Jenkins does.

**Comment by AlecWhite** [comment]
I see that the issue is https://github.com/CleverRaven/Cataclysm-DDA/issues/19586.

On an unrelated note, I'm doing what I think must be one of the biggest git fuck ups ever.

**Comment by narc0tiq** [comment]
Sorry to be the bearer of bad news, but 55abafb didn't fix it. On the other hand, the commit I just pull requested to you at AlecWhite#1 should fix the linter and actually let you make progress.

**Comment by AlecWhite** [comment]
Thanks a lot for your help.

**Comment by narc0tiq** [comment]
There we go, now the linter is being useful (see <http://gorgon.narc.ro:8080/job/Cataclysm-PR-Builder/443/console>) and recommending a specific shape for that block, which should be basically copy/paste material from here.

**Comment by Leland** [comment]
@AlecWhite you're doing great on this PR – thought I could give you a tip on how to make prettier commits, however. Git has the ability to "squash" multiple commits into one single one. See [Github's own great article](https://github.com/blog/2141-squash-your-commits) on this.

So, a workflow could be that you are locally making changes, and creating commits. Prior to pushing your new commits to your fork (and thus to this PR), you can squash any related or duplicate commits to make the history neater and more easily understood by reviewers.

It's a small thing, but it can help tremendously. Again, you're doing very well here. If you think you've made fuckups on this PR you should see some of the horrendous rebasing that's happened on past PRs.

**Comment by narc0tiq** [comment]
> Prior to pushing your new commits to your fork (and thus to this PR), you can squash any related or duplicate commits to make the history neater and more easily understood by reviewers.

Right, but in this case, he's using the PR builder for linting because it's the only linter that will accept the `monstergroup` with `starts` and `ends`.

**Comment by Leland** [comment]
@narc0tiq aah, understood. Tough break.

File: data/json/mapgen/dairy_farm.json [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    "name": "GROUP_COWS",
+    "type": "monstergroup",
+    "default": "mon_null",
+    "is_safe": true,
+    "monsters": [
+      { "monster": "mon_cow", "freq": 100, "cost_multiplier": 10, "ends": 168 },
+      { "monster": "mon_cow", "freq": 50, "cost_multiplier": 5, "starts": 336 }
+    ]
+   {,
+  {
```
Comment (narc0tiq) [inline data/json/mapgen/dairy_farm.json]: I'm surprised the linter is letting this one go -- there's an open brace, comma, and misaligned new open brace here. Maybe it's just github being weird with display.

Comment (AlecWhite) [inline data/json/mapgen/dairy_farm.json]: It isn't github being weird, it was an error.

**Comment by narc0tiq** [comment]
All right, I think we've got it now. You can squash the commits again nicely, if you wish.

**Comment by AlecWhite** [comment]
> Git has the ability to "squash" multiple commits into one single one. See this great article on this.

Thanks, didn't know you could do that.

And thanks a lot for your help @narc0tiq really appreciate it.

**Pull request merged**